docs/man/dwi2step.1
.TH dwi2step 1 "January 16th, 2003"
.SH NAME
dwi2step \- convert DWI files to pyDDR step files
.br
msd2step \- convert MSD files to pyDDR stepfiles
.br
step2dwi \- convert step files files to DWI files
.SH SYNOPSIS
\fBdwi2step \fIfilename.dwi\fR
.br
\fBmsd2step \fIfilename.msd\fR
.br
\fBstep2dwi \fIfilename.step\fR
.SH DESCRIPTION
These scripts convert the popular DWI ("Dance With Intensity", a Windows
DDR simulator) and MSD file formats into the step format used by pyDDR, and
back again. It is capable of converting all the various note types (4th,
8th, 16th, 24th,  32nd, and 64th), as well as freeze arrows (also known as
hold arrows).
.PP
After running this script, manual tweaking of the offset field in the
file may still be necessary.
.SH BUGS
BPM changes are performed at your own risk.
.PP
Certain features in step files are not present in DWI files, and so can't
be converted back in the \fBstep2dwi\fR process.
.PP
All step data for each difficulty must be in the same line for dwi2step to
convert it. Many DWI files require manual editing for this to be the case.
.SH AUTHOR
Brendan Becker <tgz@orotech.net> is the primary author of pyDDR. Joe
Wreschnig <piman@debian.org> is the maintainer of pyDDR for Debian and the
author of this manual page.
.SH SEE ALSO
\fBpyddr\fR(6)


ChangeLog
pydance ChangeLog

legend:
+ new feature added or made an existing one better
- removed a feature, or something bad happened
* fixed a bug

----- - CURRENT
0.7.5 - That's the second largest ChangeLog I've ever seen!	[2003.XX.XX]
+ New options: "Wide" to add jumps, holds on/off.
+ 5 panel (Pump It Up) support (single, versus/unison, couple, and double).
 + Support for KSF, SM, and .dance files with 5 panel steps.
 - Actual PIU pads may or may not work; none were available for testing.
+ Versus, couple, and double 6 and 8 panel modes.
+ New game type selector, so you can do e.g. Double/Endless.
+ A document on the KSF file format has been added in docs/.
+ The .dance specification has been updated with new modes and corrections.
+ A new centered display for single player modes.
+ New theming code.
 + Brings back oldschool and bryan themes.
 + Autodetect what game modes a theme will work for.
+ New 32x32 arrow set for 8/9 panel modes, by xsdg.
* Support the Thrustmaster USB mats (probably), and X-Box controllers.
 + We suspect that pydance will run on GNU/Linux on an X-Box now.
* Correctly encode/decode KSF, SM, DWI, and .dance file metadata.
 * KSF requires the Python Koco codecs package.
* DWI parsing for braindead constructs like "2!(2".
* Arrows now always draw in order (earlier under later).
* Many crashes fixed:
 * Stops + holds works again.
 * Oni lifebar works again.
 * Obscure edge case in 2 player endless mode works correctly.
* Parsing for all files is now more informative and less error-prone.
- Assist mode is broken for non-4 panel modes, we need more sounds.
- any2dance removed, it won't work anymore with this version.
- In-game menu help removed, due to inaccuracies.

0.7.4 - In My Pants						[2003.07.28]
+ 6 panel ("solo") and 8 panel support.
+ Double (1 player, both sides) support.
 - 'totaljudgings' option no longer works (no one was using it anyway...)
* Song selector UI tweaks (faster animations, nicer folder opening).
* Respace the arrows to be closer together in single/versus.
* All 9 directions can always be done on the numeric pad for player 1.
 - azerty 2nd player start key changed to '+' on the numeric pad.
* "Escape" changed from left+start to start+select.
* Some more speed optimizations for slower processors.
* Some DWI parser fixes for bad files.
* 'usepsyco' option (on by default) to try and load the Psyco JIT compiler.
- 'oldschool' and 'bryan' were removed (temporarily, maybe).

0.7.3
* Skipped, because the Linux Game Tome called 0.7.2 0.7.3 by mistake.

0.7.2 -	Dilaudid, how I love thee                               [2003.07.14]
+ Boost, brake, left, right, mirror, shuffle, and random ("supershuffle").
+ "Suck" lifebar mode, you fail if you do too good.
+ "Drop" lifebar mode, goes down at the regular speed but never up.
+ Folders in the song selector (needs > 60 songs).
* Parser bug fixes (especially for SM files).
* pyDDR has been renamed to pydance.
	
0.7.1 - OUCH MY BLADDER                                         [2003.07.01]
+ Support for COUPLE mode.
 + And battle mode.
 - any2dance doesn't always convert COUPLE from DWI to .dance correctly.
+ New player options screen (hold down start in the song selector).
+ New game options screen (hold down select).
+ "Ready? Go!" screen before the song starts.
+ New "oldschool" theme similar to earlier mixes.
* Standard round of bug fixes and tweaks.

0.7.0 - Applesauce makes the apples go yonder                   [2003.06.04]
+ More optimizations from JZig (movies this time)
+ New file formats
  + Native DWI and SM support
  + ".dance" instead of ".step" is now pyDDR's official format.
  + 'any2dance.py' converter for files.
* Judge rewrite!
  * Proper speed change handling in Judge
  * Proper miss handling in Judge

0.6.5 - Boy, did I mess that up.				[2003.05.30]
+ Major graphics optimization patch (1.5x-2x increase in fps) from JZig.
* Two player mode works again.

0.6.4 - RTFM. RTFM. RTFM. That means Read the Fucking Manual.	[2003.05.29]
+ New step file parsers
  + Beginner and hardcore (smaniac/oni/challenge) difficulty.
+ New hold arrow display
- 'mixerclock' option removed (it's now the only timing method).
- 'transcolor' option removed in favor of new 'lyriccolor' syntax.
* Bug fixes, optimizations, whippings, etc.
  + Most notably, holds, bpm changes, and stops are working much nicer now.
    - But may require reconversion with dwi2step.

0.6.3 - Reasonable Starting Times!				[2003.05.17]
+ New song loading display and algorithm (at least 4-5 times faster).
+ New song selector based roughly on DDR 5th Mix.
  + Nonstop help and playlist display.
  + 'gratuitous' option to turn off the slow extras in the song selector.
+ Endless mode.
+ Configurable oni life meter.
+ New "centered" arrow scrolling style for gameplay.
+ Smaller download size by converting wavs to oggs.
  - Requires SDL_mixer 1.2.5.
+ The timer display is controllable via the 'fpsdisplay' option.
+ New theme ('blue') and a new life display for all themes.
* DDR Extreme (8th Mix) grading and scoring algorithm.
* Split off the ddrmat.o driver from the main package.
* The constant cleanup/speedup/refactor/bug fix process.

0.6.2 - No More Ugly-Ass Buttons				[2003.04.28]
- Fullscreen switching during the game doesn't work on MacOS X.
  - Use the configuration file instead, for now.
- The pN_keys option is gone in favor of 'keyboard', and the 'sortpersist'
  option was removed.
+ 'killsongonfail' was renamed to 'autofail' and made default.
+ New game menu display.
+ Better theme detection and configuration support.
+ The 'badknees' option will remove jumps.
+ A new README that actually documents all the options!
* Once again, numerous cleanups and bugs fixed.

0.6.1 - The second one						[2003.04.09]
+ Experimental install system
  + Currently only works on Windows (and doesn't do much) and UNIX.
  + Manual pages included with pyDDR now.
  + New directory layout
+ Major code cleanups and bug fixes.

dwi2step.py
#! /usr/bin/env python

# dwi2step - converts DWI files from other clones to STEP files.
# this could peobably be done way better than I'm doing it, but..

import os, pygame, pygame.font, pygame.image, pygame.mixer, pygame.movie, time, sys
from pygame.locals import *
from Numeric import *

VERSION = "0.51"

# so it's currently in one routine. shut up, I'm learning python =]
def main():
  print "dwi2step, by theGREENzebra (tgz@clickass.org)"

  # set up
  print "Initialising.."

# ARROW STATUSES in arrstat
#-:   0 = no status / not initialised / top arrow
#0:   1 = shown
#1:   2 = pulse
#2:   4 = step
#3:   8 = onbeat
#4:  16 = offbeat
#5:  32 = way offbeat
#6:  64 =
#7: 128 = freeze

  # If I knew how to check whether a variable was defined, I would do
  # that instead
  bpm = ''
  gap = ''
  stepmodes = ['sixty','twtfr','steps','eight','qurtr','halfn','whole']
  stepbeats = [ 1/64.0, 1/24.0,     1 ,     2 ,     4 ,     8 ,    16 ]
  rar = 0
  bpms = []
  delay_times = []
  bpmchange_beats = []
  delay_beats = []

#  arrstat = zeros((4,1024,3))
  print sys.argv[0]

  if len(sys.argv) > 1:
    print "Opening DWI file:",sys.argv[1],".."
    msdf = sys.argv[1]
    if msdf[-4:] != ".dwi": 
      if msdf[-4:] != ".DWI": 
        msdf += ".dwi"
    msdfile = open(msdf)
    fileline = 'crapcrap'

    stepfilename = msdf
    stepfilename = stepfilename[:-4]+".step"
    print 'Opening STEP file:',stepfilename,'..'
    stepfile = open(stepfilename,"w") #but for writing
    stepfile.write("version " + VERSION + "\n")
    print "debugging crap follows"
    print "----------------------"
    print "wrote header"
  else:
    print "You need to actually specify an DWI file to convert!"
    fileline = ''

  while fileline != '':
    print '*',
    if fileline != '\n':
      fileline = fileline.strip()
      if fileline[0:6] == '#FILE:':
        songfile = fileline[6:(len(fileline)-5)]
        print "found song filename: ", songfile
        stepfile.writelines("file "+songfile.lower()+".mp3\n")
      if fileline[0:7] == '#TITLE:':
        songname = fileline[7:(len(fileline)-1)]
        print "found songname: ", songname
        stepfile.writelines("song "+songname+"\n")
      if fileline[0:5] == '#BPM:':
        bpm = float(fileline[5:(len(fileline)-1)])
        print "found bpm: ", bpm
        stepfile.writelines("bpm "+"%.2f"%bpm+"\n")
      if fileline[0:11] == '#CHANGEBPM:':
        parsemode = 1
        beat_string = ''
        bpmchange_string = ''
        print "found bpm changes: "
        for i in fileline[11:-1]:
          if i == '=':   
            bpmchange_beats.append(float(beat_string))
            parsemode = 2
            bpmchange_string = ''
          elif i == ',':
            bpms.append(float(bpmchange_string))
            parsemode = 1
            beat_string = ''
          else:
            if parsemode==1:
              beat_string += i
            if parsemode==2:
              bpmchange_string += i
        bpms.append(float(bpmchange_string))
        for i in bpmchange_beats:
          print "  at beat",i,"change bpm to",bpms[bpmchange_beats.index(i)]
        print
      if fileline[0:8] == '#FREEZE:':
        parsemode = 1
        beat_string = ''
        delay_string = ''
        print "frozen beats: "
        for i in fileline[8:-1]:
          if i == '=':
            delay_beats.append(float(beat_string))
            parsemode = 2
            delay_string = ''
          elif i == ',':
            delay_times.append(float(delay_string))
            parsemode = 1
            beat_string = ''
          else:
            if parsemode==1:
              beat_string += i
            if parsemode==2:
              delay_string += i
        delay_times.append(float(delay_string))
        for i in delay_beats:
          print "  at beat",i,"freeze",delay_times[delay_beats.index(i)]
        print
      if fileline[0:8] == '#ARTIST:':
        groupname = fileline[8:(len(fileline)-1)]
        print "found groupname: ", groupname
        stepfile.writelines("group "+groupname+"\n")
      if fileline[0:5] == '#GAP:':
        gap = int(fileline[5:-1])
        print "found gap: ", gap, "( => offset = ", -gap, ")"
        stepfile.writelines("offset "+repr(-gap)+"\n")
      if fileline[0:14] == '#SINGLE:BASIC:':
        onbeat = 0
        diff = int(fileline[14])
        print "\ndifficulty SINGLE BASIC,",diff,"feet. reading steps.."
        stepfile.writelines("SINGLE\n"+"BASIC "+repr(diff)+"\n")
        fileline = fileline[16:] + "\n"
        dc = ''
        stepmode = "eight"
        rar = 1
      if fileline[0:14] == '#SINGLE:TRICK:':
        onbeat = 0
        diff = int(fileline[14])
        print "\ndifficulty SINGLE TRICK,",diff,"feet. reading steps.."
        stepfile.writelines("SINGLE\n"+"TRICK "+repr(diff)+"\n")
        fileline = fileline[16:] + "\n"
        dc = ''
        stepmode = "eight"
        rar = 1
      if fileline[0:15] == '#SINGLE:MANIAC:':
        onbeat = 0
        diff = int(fileline[15])
        print "\ndifficulty SINGLE MANIAC,",diff,"feet. reading steps.."
        stepfile.writelines("SINGLE\n"+"MANIAC "+repr(diff)+"\n")
        fileline = fileline[17:] + "\n"
        dc = ''
        stepmode = "eight"
        rar = 1
      if fileline[0:16] == '#SINGLE:ANOTHER:':
        onbeat = 0
        diff = int(fileline[16])
        print "\ndifficulty SINGLE TRICK,",diff,"feet. reading steps.."
        stepfile.writelines("SINGLE\n"+"TRICK "+repr(diff)+"\n")
        fileline = fileline[18:] + "\n"
        dc = ''
        stepmode = "eight"
        rar = 1
      
      i = 0
      while rar:
        steps = []

        for xyz in bpmchange_beats:    #check to see if we need to change the bpm
          if xyz == onbeat:
            stepfile.writelines("chbpm "+repr(bpms[bpmchange_beats.index(onbeat)])+"\n")
            print "-- chbpm to",bpms[bpmchange_beats.index(onbeat)],"at beat",xyz,"--",
        for xyz in delay_beats:    #check to see if we need to wait some time
          if xyz == onbeat:
            stepfile.writelines("waits "+repr(delay_times[delay_beats.index(onbeat)]/1000)+"\n")
            print "-- dalay of",delay_times[delay_beats.index(onbeat)]/1000,"sec. at beat",xyz,"--",

        dc = fileline[i]
        dc1 = dc2 = ''       #initialized here so that they are declared even on EOF/EOL
        try:
          dc1 = fileline[i+1]
          dc2 = fileline[i+2]
        except:              #EOF or EOL
          neareof = 1
          
        i += 1

        print dc,
        if dc == "0":
          steps = ['00','00','00','00']
          onbeat += stepbeats[stepmodes.index(stepmode)]
        elif dc == "1":
          steps = ['08','08','00','00']
          onbeat += stepbeats[stepmodes.index(stepmode)]
        elif dc == "2":
          steps = ['00','08','00','00']
          onbeat += stepbeats[stepmodes.index(stepmode)]
        elif dc == "3":
          steps = ['00','08','00','08']
          onbeat += stepbeats[stepmodes.index(stepmode)]
        elif dc == "4":
          steps = ['08','00','00','00']
          onbeat += stepbeats[stepmodes.index(stepmode)]
        elif dc == "5":
          steps = ['00','00','00','00']
          onbeat += stepbeats[stepmodes.index(stepmode)]
        elif dc == "6":
          steps = ['00','00','00','08']
          onbeat += stepbeats[stepmodes.index(stepmode)]
        elif dc == "7":
          steps = ['08','00','08','00']
          onbeat += stepbeats[stepmodes.index(stepmode)]
        elif dc == "8":
          steps = ['00','00','08','00']
          onbeat += stepbeats[stepmodes.index(stepmode)]
        elif dc == "9": 
          steps = ['00','00','08','08']
          onbeat += stepbeats[stepmodes.index(stepmode)]
        elif dc == "A":
          steps = ['00','08','08','00']
          onbeat += stepbeats[stepmodes.index(stepmode)]
        elif dc == "B":
          steps = ['08','00','00','08']
          onbeat += stepbeats[stepmodes.index(stepmode)]
        if dc == "(":
          stepmode = "steps"
        if dc == ")":
          stepmode = "eight"
        if dc == "[":
          stepmode = "twtfr"
        if dc == "]":
          stepmode = "eight"
        if dc == "{":
          stepmode = "sixty"
        if dc == "}":
          stepmode = "eight"
        if dc == ";":
          stepfile.writelines("end\n")
          rar = 0
        if dc == "\n":
          fileline = msdfile.readline()
          i = 0

        if (not neareof) and (dc1 == '!'):    #freeze arrow
          i += 2
          print 'hold',dc2,

          if dc2 == "1":
            steps[0] =  '8'+steps[0][1]
            steps[1] =  '8'+steps[1][1]
          elif dc2 == "2":
            steps[1] =  '8'+steps[1][1]
          elif dc2 == "3":
            steps[1] =  '8'+steps[1][1]
            steps[3] =  '8'+steps[3][1]
          elif dc2 == "4":
            steps[0] =  '8'+steps[0][1]
          elif dc2 == "6":
            steps[3] =  '8'+steps[3][1]
          elif dc2 == "7":
            steps[0] =  '8'+steps[0][1]
            steps[2] =  '8'+steps[2][1]
          elif dc2 == "8":
            steps[2] =  '8'+steps[2][1]
          elif dc2 == "9": 
            steps[2] =  '8'+steps[2][1]
            steps[3] =  '8'+steps[3][1]
          elif dc2 == "A":
            steps[1] =  '8'+steps[1][1]
            steps[2] =  '8'+steps[2][1]
          elif dc2 == "B":
            steps[0] =  '8'+steps[0][1]
            steps[3] =  '8'+steps[3][1]

        stepline = stepmode+" "
        for stepwrite in steps:    stepline += stepwrite + " "
        if steps:
#          print stepline
          stepfile.writelines(stepline+"\n")

    fileline = msdfile.readline()

  print "End."
#end

if __name__ == '__main__': main()
